src/lib/y2partitioner/dialogs/partition_size.rb

```diff
@@ -10,143 +13,355 @@ module Dialogs
-              MinWidth(
-                10,
-                IntField(
-                  Id(:start_block),
-                  _("Start Block"),
-                  min_block, max_block, start_block
-                )
-              ),
-              MinWidth(
-                10,
-                IntField(
-                  Id(:end_block),
-                  _("End Block"),
-                  min_block, max_block, end_block
-                )
+        # @macro seeAbstractWidget
+        def store
+          # nothing to do, that's OK
+        end
+      end
+
+      # Enter a human readable size
+      class CustomSizeInput < CWM::InputField
+        # @return [Y2Storage::DiskSize]
+        attr_reader :min_size, :max_size
+
+        # @param ptemplate [Sequences::PartitionTemplate]
+        #   a partition template, collecting data for a partition to be created
+        # @param regions [Array<Y2Storage::Region>]
+        #   regions available to create a partition in
+        def initialize(ptemplate, regions)
+          textdomain "storage"
+          @ptemplate = ptemplate
+          @regions = regions
+          largest_region = @regions.max_by(&:size)
+          @max_size = largest_region.size
+          @min_size = Y2Storage::DiskSize.new(1)
+        end
+
+        # Forward to ptemplate
+        def size
+          @ptemplate.custom_size
+        end
+
+        # Forward to ptemplate
+        def size=(v)
+          @ptemplate.custom_size = v
+        end
+
+        # @return [Y2Storage::Region] the smallest region
+        #   that can contain the chosen size
+        def parent_region
+          suitable_rs = @regions.find_all { |r| r.size >= size }
+          suitable_rs.min_by(&:size)
+        end
+
+        # @return [Y2Storage::Region]
+        # A new region created in the smallest region
+        #   that can contain the chosen size
+        def region
+          parent = parent_region
+          bsize = parent.block_size
+          length = (size.to_i / bsize.to_i.to_f).ceil
+          Y2Storage::Region.create(parent.start, length, bsize)
+        end
+
+        # @macro seeAbstractWidget
+        def label
+          _("Size")
+        end
+
+        # @macro seeAbstractWidget
+        def init
+          self.size ||= max_size
+          self.value = size
+        end
+
+        # @macro seeAbstractWidget
+        def store
+          self.size = value
+        end
+
+        # @macro seeAbstractWidget
+        def validate
+          v = value
+          if v.nil? || v > max_size
+            min_s = min_size.human_ceil
+            max_s = max_size.human_floor
+            Yast::Popup.Error(
+              Yast::Builtins.sformat(
+                # error popup, %1 and %2 are replaced by sizes
+                _("The size entered is invalid. Enter a size between %1 and %2."),
+                min_s, max_s
               )
             )
-          )
+            # TODO: Let CWM set the focus
```

Comment:
it is not so easy due to custom widgets, where you want usually focus specific part of that widget.